pkg/restore_snapshot.go

```diff
@@ -0,0 +1,207 @@
+package pkg
+
+import (
+	"context"
+
+	"stash.appscode.dev/apimachinery/apis/stash/v1beta1"
+	api_v1beta1 "stash.appscode.dev/apimachinery/apis/stash/v1beta1"
+	stash "stash.appscode.dev/apimachinery/client/clientset/versioned"
+	"stash.appscode.dev/apimachinery/pkg/invoker"
+	"stash.appscode.dev/apimachinery/pkg/restic"
+
+	"github.com/spf13/cobra"
+	"gomodules.xyz/flags"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	appcatalog "kmodules.xyz/custom-resources/apis/appcatalog/v1alpha1"
+	appcatalog_cs "kmodules.xyz/custom-resources/client/clientset/versioned"
+	v1 "kmodules.xyz/offshoot-api/api/v1"
+)
+
+var downloadDir, name, initialAdvertisePeerUrl, initialCluster, initialClusterToken string
+
+func NewCmdRestoreSnapshot() *cobra.Command {
+	var (
+		masterURL      string
+		kubeconfigPath string
+		opt            = etcdOptions{
+			setupOptions: restic.SetupOptions{
+				ScratchDir:  restic.DefaultScratchDir,
+				EnableCache: false,
+			},
+			waitTimeout: 300,
+			restoreOptions: restic.RestoreOptions{
+				Host: restic.DefaultHost,
+			},
+		}
+	)
+
+	cmd := &cobra.Command{
+		Use:               "restore-snapshot",
+		Short:             "Download Etcd DB backup from cloud and restore the snapshot in the restore directory",
+		DisableAutoGenTag: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			flags.EnsureRequiredFlags(cmd, "appbinding", "provider", "secret-dir", "hostname")
+
+			// prepare client
+			config, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfigPath)
+			if err != nil {
+				return err
+			}
+			opt.kubeClient, err = kubernetes.NewForConfig(config)
+			if err != nil {
+				return err
+			}
+			opt.catalogClient, err = appcatalog_cs.NewForConfig(config)
+			if err != nil {
+				return err
+			}
+			opt.stashClient, err = stash.NewForConfig(config)
+			if err != nil {
+				return err
+			}
+
+			targetRef := api_v1beta1.TargetRef{
+				APIVersion: appcatalog.SchemeGroupVersion.String(),
+				Kind:       appcatalog.ResourceKindApp,
+			}
+
+			err = opt.downloadSnapshot(targetRef)
+			if err != nil {
+				return err
+			}
+
```

Comment:
`// Now, restore the downloaded snapshot from interim directory into a sub-directory of this etcd data directory.`